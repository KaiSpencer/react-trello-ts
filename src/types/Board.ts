import {CSSProperties} from 'react'

export interface BoardData {
  lanes: Lane[]
}

export interface Lane {
  id: string
  title?: string
  label?: string
  style?: CSSProperties
  cards: Card[]
  currentPage?: number
  droppable?: boolean
  labelStyle?: CSSProperties
  cardStyle?: CSSProperties
  disallowAddingCard?: boolean
  [key: string]: any
}

export interface Card {
  id: string
  title?: string
  label?: string
  description?: string
  laneId?: string
  style?: CSSProperties
  draggable?: boolean
  [key: string]: any
}
