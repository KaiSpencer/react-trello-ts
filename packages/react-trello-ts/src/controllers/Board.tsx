import React, { createContext, FC, PropsWithChildren } from "react";
import classNames from "classnames";
import uuidv1 from "uuid/v1";
import { BoardContainer } from "./BoardContainer";
import * as DefaultComponets from "../components";
import { BoardData } from "../types/Board";
import { store } from "../store/store";

export const Board: FC<
	PropsWithChildren<{
		id?: string;
		className?: string;
		components: typeof DefaultComponets;
		data: BoardData;
		t?: any;
	}>
> = ({ data, children, className, components, id, t, ...rest }) => {
	const allClassNames = classNames("react-trello-board", className || "");

	return (
		<BoardContext.Provider value={store}>
			<components.GlobalStyle />
			<BoardContainer
				id={id || uuidv1()}
				data={data}
				components={components}
				t={t}
				className={allClassNames}
				{...rest}
			/>
		</BoardContext.Provider>
	);
};

export const BoardContext = createContext(store);
