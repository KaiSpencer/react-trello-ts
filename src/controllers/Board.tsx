import classNames from "classnames";
import React, { FC, PropsWithChildren, createContext } from "react";
import { v1 } from "uuid";
import * as DefaultComponets from "../components";
import { store } from "../store/store";
import { BoardData } from "../types/Board";
import { BoardContainer } from "./BoardContainer";

export const Board: FC<
	PropsWithChildren<{
		id?: string;
		className?: string;
		components: typeof DefaultComponets;
		data: BoardData;
		t?: any;
	}>
> = ({ data, children, className, components, id, t, ...rest }) => {
	const allClassNames = classNames("react-trello-board", className || "");

	return (
		<BoardContext.Provider value={store}>
			{/*
			// @ts-ignore */}
			<components.GlobalStyle />
			<BoardContainer
				id={id || v1()}
				data={data}
				components={components}
				t={t}
				className={allClassNames}
				{...rest}
			/>
		</BoardContext.Provider>
	);
};

export const BoardContext = createContext(store);
