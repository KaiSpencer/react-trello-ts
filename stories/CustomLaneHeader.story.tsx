import React from 'react'
import {storiesOf} from '@storybook/react'

import Board from '../src'

const CustomLaneHeader = ({label, cards, title, current, target}) => {
  const buttonHandler = () => {
    alert(`The label passed to the lane was: ${label}. The lane has ${cards.length} cards!`)
  }
  return (
    <div>
      <header
        style={{
          borderBottom: '2px solid #c5c5c5',
          paddingBottom: 6,
          marginBottom: 10,
          display: 'flex',
          flexDirection: 'row',
          justifyContent: 'space-between'
        }}>
        <div style={{fontSize: 14, fontWeight: 'bold'}}>{title}</div>
        {label && (
          <div style={{width: '30%', textAlign: 'right', fontSize: 13}}>
            <button onClick={buttonHandler} style={{cursor: 'pointer'}}>
              ?
            </button>
          </div>
        )}
      </header>
      <div>
        Percentage: {current || 0}/{target}
      </div>
    </div>
  )
}

storiesOf('Custom Components', module).add(
  'LaneHeader',
  () => {
    const data = {
      lanes: [
        {
          id: 'lane1',
          title: 'Planned Tasks',
          current: '70', // custom property
          target: '100', // custom property
          label: 'First Lane here',
          cards: [
            {
              id: 'Card1',
              title: 'John Smith',
              description: 'Thanks. Please schedule me for an estimate on Monday.'
            },
            {
              id: 'Card2',
              title: 'Card Weathers',
              description: 'Email received at 1:14pm'
            }
          ]
        },
        {
          id: 'lane2',
          title: 'Completed Tasks',
          label: 'Second Lane here',
          current: '30', // custom property
          target: '100', // custom property
          cards: [
            {
              id: 'Card3',
              title: 'Michael Caine',
              description: 'You are welcome. Interested in doing business with you' + ' again',
              tags: [
                {title: 'Critical', color: 'white', bgcolor: 'red'},
                {title: '2d ETA', color: 'white', bgcolor: '#0079BF'}
              ]
            }
          ]
        }
      ]
    }

    return <Board data={data} components={{LaneHeader: CustomLaneHeader}} />
  },
  {info: 'Style your lane header appearance'}
)
