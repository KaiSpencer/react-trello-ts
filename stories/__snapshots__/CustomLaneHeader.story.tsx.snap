// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Components LaneHeader smoke-test 1`] = `
<div draggable="false"
     class="sc-iGgWBj iSNfA"
>
  <div>
    <div class="smooth-dnd-container horizontal">
      <section target="100"
               draggable="false"
               class="sc-kAyceB dgWEjS react-trello-lane"
      >
        <div>
          <header style="border-bottom: 2px solid rgb(197, 197, 197); padding-bottom: 6px; margin-bottom: 10px; display: flex; flex-direction: row; justify-content: space-between;">
            <div style="font-size: 14px; font-weight: bold;">
              Planned Tasks
            </div>
            <div style="width: 30%; text-align: right; font-size: 13px;">
              <button type="submit"
                      style="cursor: pointer;"
              >
                ?
              </button>
            </div>
          </header>
          <div>
            Percentage: 0/
          </div>
        </div>
        <div class="sc-dhKdcB ikKiDc">
          <div class="smooth-dnd-container vertical">
            <div class="smooth-dnd-draggable-wrapper">
              <article data-id="Card1"
                       class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
              >
                <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                  <span draggable="true"
                        class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                  >
                    John Smith
                  </span>
                  <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                  </span>
                  <div class="sc-hmdomO gJuJJJ">
                    <button class="sc-jsJBEP djrAfI">
                      ✖
                    </button>
                  </div>
                </header>
                <div class="sc-eDPEul cRhiim">
                  Thanks. Please schedule me for an estimate on Monday.
                </div>
              </article>
            </div>
            <div class="smooth-dnd-draggable-wrapper">
              <article data-id="Card2"
                       class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
              >
                <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                  <span draggable="true"
                        class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                  >
                    Card Weathers
                  </span>
                  <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                  </span>
                  <div class="sc-hmdomO gJuJJJ">
                    <button class="sc-jsJBEP djrAfI">
                      ✖
                    </button>
                  </div>
                </header>
                <div class="sc-eDPEul cRhiim">
                  Email received at 1:14pm
                </div>
              </article>
            </div>
          </div>
        </div>
      </section>
      <section target="100"
               draggable="false"
               class="sc-kAyceB dgWEjS react-trello-lane"
      >
        <div>
          <header style="border-bottom: 2px solid rgb(197, 197, 197); padding-bottom: 6px; margin-bottom: 10px; display: flex; flex-direction: row; justify-content: space-between;">
            <div style="font-size: 14px; font-weight: bold;">
              Completed Tasks
            </div>
            <div style="width: 30%; text-align: right; font-size: 13px;">
              <button type="submit"
                      style="cursor: pointer;"
              >
                ?
              </button>
            </div>
          </header>
          <div>
            Percentage: 0/
          </div>
        </div>
        <div class="sc-dhKdcB ikKiDc">
          <div class="smooth-dnd-container vertical">
            <div class="smooth-dnd-draggable-wrapper">
              <article data-id="Card3"
                       class="sc-jlZhew sc-cwHptR clGdKf fQNcVl react-trello-card"
              >
                <header class="sc-gsFSXq sc-jEACwC imRIDz jNaWkr">
                  <span draggable="true"
                        class="sc-kpDqfm sc-cPiKLX cFicSq cnDDQK"
                  >
                    Michael Caine
                  </span>
                  <span class="sc-dAlyuH sc-dLMFU iHiWje dvkFIZ">
                  </span>
                  <div class="sc-hmdomO gJuJJJ">
                    <button class="sc-jsJBEP djrAfI">
                      ✖
                    </button>
                  </div>
                </header>
                <div class="sc-eDPEul cRhiim">
                  You are welcome. Interested in doing business with you again
                </div>
                <div class="sc-eldPxv iDFVov">
                  <span class="sc-fPXMVe eCIUDc"
                        style="color: white; background-color: red;"
                  >
                    Critical
                  </span>
                  <span class="sc-fPXMVe eCIUDc"
                        style="color: white; background-color: rgb(0, 121, 191);"
                  >
                    2d ETA
                  </span>
                </div>
              </article>
            </div>
          </div>
        </div>
      </section>
    </div>
  </div>
</div>
`;
